### PR TITLE
Feature/#3 grammar

### PR DESCRIPTION
**Pull request by jetmartin**
Merging [Syntax highlight](https://github.com/jetmartin/bats/issues/3) updates.

The file has been updated to provide syntax highlight for [bats-core](https://github.com/bats-core/bats-core), [bats-support](https://github.com/ztombol/bats-support) & [bats-assert](https://github.com/ztombol/bats-assert).

The README has also been updated.

@Flamefire, thanks for your feedback on that one.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,7 +43,12 @@
                 "language": "bats",
                 "scopeName": "source.shell.bats",
                 "path": "./syntaxes/bats.tmLanguage"
-            }
+            },
+			{
```
Comment (Flamefire) [inline package.json]: indendation

File: syntaxes/bats.extended.tmLanguage [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            <!-- bats -->
+            <dict>
+                <key>match</key>
+                <string>\b(run|load|skip)\b</string>
+                <key>name</key>
+                <string>support.function.bats.builtin</string>
+            </dict>
+            <dict>
+                <key>match</key>
+                <string>\$(output|status|lines)\b</string>
```
Comment (Flamefire) [inline syntaxes/bats.extended.tmLanguage]: `$lines` is usually used as `"${lines[0]}", ${lines[0]}` or `"${lines[@]}`/ `"${lines[*]}` Could you add those as well?

I'm most worried about the quoted form, because `injectionSelector` excludes strings. Probably need another file...

File: test/test.bats [hunk 3]
```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env bats
```
Comment (Flamefire) [inline test/test.bats:1]: What is the purpose of this file? It seems to contain lot's of test code of external libs which seems out of scope of this lib

Comment (jetmartin) [reply test/test.bats:1]: That's just a test file I've used to check the highlights of each functions.
The code came from the libs repositories (copied from test files for each function).
I didn't find an easy way to automate grammar coloration testing.
Not mandatory to keep in sources.

Comment (Flamefire) [reply test/test.bats:1]: Then trim it down:
- Use every/most functions that should be highlighted. Only syntactically correct (no need for valid tests, keep it short)
- Use (some) functions in strings (in code and test name) and comments
- Use variables
- Opt: Add comments what to look for

File: syntaxes/bats.extended.tmLanguage (hunk 2 above)
Comment (jetmartin) [reply syntaxes/bats.extended.tmLanguage]: I will test how it is managed by VSC. I din't get time yesterday.
I guess the wildcard is the simplest solution if it's well managed by VSC.

Comment (jetmartin) [reply syntaxes/bats.extended.tmLanguage]: Just thinking about it.
The variables are already coloured by SH syntax without the `support.variable.bats` element.
I don't see the use-case for that. Could you provide an exemple ?
Screenshot 2019-06-20 at 13 59 40

Comment (Flamefire) [reply syntaxes/bats.extended.tmLanguage]: the variables are colored as shell variables. But they have special meaning in bats. So I'd like to have `$output` colored differently than `$my_output`

Use cases:
- Easier visible
- Detects typos like `$uotput`

File: test/test.bats (hunk 3 above)
Comment (jetmartin) [reply test/test.bats:1]: You'r right, I was a bit quick and dirty.
I've jsut push an updated file.

File: syntaxes/bats.extended.tmLanguage (hunk 2 above)
Comment (jetmartin) [reply syntaxes/bats.extended.tmLanguage]: I understand the use-cases but I'm not completely convinced for two reasons :

First, Bats is using a valid sh syntax.
The tree bats variables are just "reserved" variables names used by BATS.
So if I use shellcheck, the bats file should be valid (the green one are not on my screenshot, see next point).

Second, I've found no satisfying way to be able to manage the colouring if you use quotes or with any kind of brackets. (and I've tested many regex)
And that's logic so far as bats syntax is build on top of shell one, the regex is just apply on vars not on vars parameters or strings.
To manage that may should we create a new language definition... may actuals suff is enough so far ?
See bellow, just the two first ones are well coloured :
image

Let me knew what you think or if you have additional informations on that topic.
Otherwise, we could probably close it or may this could stay as a new open issue requiring additional support.

Best.

Comment (Flamefire) [reply syntaxes/bats.extended.tmLanguage]: > The tree bats variables are just "reserved" variables names used by BATS.
So if I use shellcheck, the bats file should be valid (the green one are not on my screenshot, see next point).

Agreed. But why is this against this proposal? Highlighting the special variables is useful because they are special. Just the same as the functions.

I was able to find a way: A 2nd language injection. Remove the variables from the first (source.shell.bats.extended), then add another entry to package.json

                      {
				"scopeName": "source.shell.bats.variables",
				"path": "./syntaxes/bats.variables.tmLanguage",
				"injectTo": [ "source.shell.bats" ]
			}

And finally the new definition:

```
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>scopeName</key>
        <string>source.shell.bats.variables</string>
        <key>injectionSelector</key>
        <string>L:source.shell.bats -comment</string>
        <key>patterns</key>
        <array>
            <dict>
                <key>match</key>
                <string>\$(output|status|lines)\b</string>
                <key>name</key>
                <string>support.variable.bats</string>
            </dict>
            <dict>
                <key>match</key>
                <string>\$\{#?(output|status|lines)\b</string>
                <key>captures</key>
                <dict>
                    <key>1</key>
                    <dict>
                        <key>name</key>
                        <string>support.variable.bats</string>
                    </dict>
                </dict>
            </dict>
        </array>
    </dict>
</plist>
```

But I don't insist on that. Would be nice to have, but having the functions already helps

Comment (jetmartin) [reply syntaxes/bats.extended.tmLanguage]: Hi,

Well done, I didn't think about it !
It's not perfect but works.
image

So I've just push an update where I allow colour change on settings and I provided a sample setting in README.

Could you please test the updated module on your real files before I merge ?
Thanks

**Review by Flamefire — APPROVED** [review]

File: syntaxes/bats.extended.tmLanguage (hunk 2 above)
Comment (Flamefire) [reply syntaxes/bats.extended.tmLanguage]: Great! :)

**Comment by Flamefire** [comment]
LGTM. last change: I'd add `stub` and `unstub` from bats-mock to the functions

**Comment by jetmartin** [comment]
I've added :
 - setup and teardown functions who were missing
 - bats-mock

Please confirm it's working fine on your files before I merge.

**Comment by Flamefire** [comment]
Please add them to a different group. The rest highlights calls while those 2 are rather meant for definitions. You don't call them. So please use an extra category like `support.function.bats.framework` or so

**Comment by jetmartin** [comment]
ok, I'll just rollbak that so far.
